toolsrc/src/vcpkg/base/downloads.cpp

```diff
@@ -167,30 +201,257 @@ namespace vcpkg::Downloads
+            {
+                Checks::unreachable(VCPKG_LINE_INFO);
+            }
+
+            // Make sure the directories are present, otherwise fopen_s fails
+            const auto dir = download_path_part_path.parent_path();
+            fs.create_directories(dir, VCPKG_LINE_INFO);
+
+            WriteFlushFile f(download_path_part_path);
+
+            auto path_begin = Util::find(url_no_proto, '/');
+            std::string hostname(url_no_proto.begin(), path_begin);
+            std::string path(path_begin, url_no_proto.end());
+
+            Debug::print("Downloading ", url, "\n");
+            static auto s = WinHttpSession::make().value_or_exit(VCPKG_LINE_INFO);
+            auto conn = WinHttpConnection::make(s.m_hSession.get(), hostname, port);
+            if (!conn)
+            {
+                Strings::append(errors, url, ": ", conn.error(), '\n');
+                return false;
+            }
+            auto req = WinHttpRequest::make(conn.get()->m_hConnect.get(), path);
+            if (!req)
+            {
+                Strings::append(errors, url, ": ", req.error(), '\n');
+                return false;
+            }
+            auto forall_data =
+                req.get()->forall_data([&](Span<char> span) { fwrite(span.data(), 1, span.size(), f.f); });
+            if (!forall_data)
+            {
+                Strings::append(errors, url, ": ", forall_data.error(), '\n');
+                return false;
+            }
+            return true;
+        }
+    }
+#endif
+
+    std::string download_file(vcpkg::Files::Filesystem& fs,
+                              View<std::string> urls,
+                              const fs::path& download_path,
+                              const std::string& sha512)
+    {
+        Checks::check_exit(VCPKG_LINE_INFO, urls.size() > 0);
+
+        auto download_path_part_path = download_path;
+        download_path_part_path += fs::u8path(".part");
+        fs.remove(download_path, ignore_errors);
+        fs.remove(download_path_part_path, ignore_errors);
+
+        std::string errors;
+        for (const std::string& url : urls)
+        {
 #if defined(_WIN32)
-        auto url_no_proto = url.substr(8); // drop https://
-        auto path_begin = Util::find(url_no_proto, '/');
-        std::string hostname(url_no_proto.begin(), path_begin);
-        std::string path(path_begin, url_no_proto.end());
-
-        winhttp_download_file(fs, download_path_part, hostname, path);
-#else
-        const auto code = System::cmd_execute(
-            Strings::format(R"(curl -L '%s' --create-dirs --output '%s')", url, download_path_part));
-        Checks::check_exit(VCPKG_LINE_INFO, code == 0, "Could not download %s", url);
+            if (Strings::starts_with(url, "https://") || Strings::starts_with(url, "http://"))
+            {
+                StringView without_proto(url);
+                without_proto = url[4] == 's' ? without_proto.substr(8) : without_proto.substr(7);
```

Comment:
```suggestion
                without_proto = without_proto.substr(url[4] == 's' ? 8 : 7);
```

(Or maybe even `without_proto = without_proto.substr(7 + (url[4] == 's'));`